backend/hct_mis_api/apps/program/api/caches.py

```diff
@@ -0,0 +1,25 @@
+from typing import Any
+
+from rest_framework_extensions.key_constructor.bits import KeyBitBase
+
+from hct_mis_api.api.caches import KeyConstructorMixin, ProgramKeyBit
+from hct_mis_api.apps.core.utils import decode_id_string
+from hct_mis_api.apps.program.models import ProgramCycle
+
+
+class ProgramCycleListVersionsKeyBit(KeyBitBase):
+    def get_data(
+        self, params: Any, view_instance: Any, view_method: Any, request: Any, args: tuple, kwargs: dict
+    ) -> str:
+        program_id = decode_id_string(kwargs.get("program_id"))
+        program_cycle_updated_at = (
+            ProgramCycle.all_objects.filter(program_id=program_id).latest("updated_at").updated_at
+        )
+
+        version_key = f"{program_id}:{program_cycle_updated_at}"
```

Comment:
Although this is a nice idea, it has cases where this approach would fail - for example removing ProgramCycles (not the newest ones), will leave us with the same cache version.

Reply:
we are using soft delete in ProgramCycle, so after removing ProgramCycle updated_at will also be changed to the current date, that's why we are using `all_objects` manager to fetch also soft deleted objects